src/haxeLanguageServer/features/ExtractTypeFeature.hx

```diff
@@ -0,0 +1,148 @@
+package haxeLanguageServer.features;
+
+import haxe.io.Path;
+import sys.FileSystem;
+import haxeLanguageServer.helper.WorkspaceEditHelper;
+import tokentree.TokenTree;
+import tokentree.utils.TokenTreeCheckUtils;
+
+using tokentree.TokenTreeAccessHelper;
+
+class ExtractTypeFeature {
+	final context:Context;
+
+	public function new(context:Context) {
+		this.context = context;
+		#if debug
+		context.registerCodeActionContributor(extractType);
+		#end
+	}
+
+	function extractType(params:CodeActionParams):Array<CodeAction> {
+		var doc = context.documents.get(params.textDocument.uri);
+		try {
+			var fsPath:FsPath = params.textDocument.uri.toFsPath();
+			var path:Path = new Path(fsPath.toString());
+			if (path.ext != "hx")
+				return [];
+
+			if ((doc.tokens == null) || (doc.tokens.tree == null))
+				return [];
+
+			var types:Array<TokenTree> = doc.tokens.tree.filterCallback(function(token:TokenTree, index:Int):FilterResult {
+				switch (token.tok) {
+					case Kwd(KwdClass), Kwd(KwdInterface), Kwd(KwdEnum), Kwd(KwdAbstract), Kwd(KwdTypedef):
+						return FOUND_SKIP_SUBTREE;
+					default:
+				}
+				return GO_DEEPER;
+			});
+			var lastImport:Null<TokenTree> = getLastImportToken(doc);
+			if (isInsideConditional(lastImport))
+				return [];
+
+			var fileHeader = "";
+
+			// copy all imports from current file
```

Comment:
When I tried this for the first time, I got compilation errors even though it copies all imports. That's because it doesn't add an import to the module it was extracted from (in my case it depended on one of the other subtypes there).

Reply:
yeah, it probably should collect all identifiers of extracted type and compare them with types remaining in file, so that it can handle such cases